protocol/pubsub/v2/internal/connection_test.go

```diff
@@ -161,6 +161,58 @@ func TestPublishCreateTopic(t *testing.T) {
 	verifyTopicDeleteWorks(t, client, psconn, topicID)
 }
 
+// Test that publishing to a topic with non default publish settings
+func TestPublishWithCustomPublishSettings(t *testing.T) {
+	t.Run("create topic and publish to it with custom settings", func(t *testing.T) {
+		ctx := context.Background()
+		pc := &testPubsubClient{}
+		defer pc.Close()
+
+		projectID, topicID, subID := "test-project", "test-topic", "test-sub"
+
+		client, err := pc.New(ctx, projectID, nil)
+		if err != nil {
+			t.Fatalf("failed to create pubsub client: %v", err)
+		}
+		defer client.Close()
+
+		psconn := &Connection{
+			AllowCreateSubscription: true,
+			AllowCreateTopic:        true,
+			Client:                  client,
+			ProjectID:               projectID,
+			TopicID:                 topicID,
+			SubscriptionID:          subID,
+			PublishSettings: &pubsub.PublishSettings{
```

Comment:
The change itself looks good to me, but I'm curious how this tests it. What in the test makes it fail and what would change if the PublishSettings were not specified?

Reply:
If not specified then default publish settings would be used as documented here https://github.com/googleapis/google-cloud-go/blob/265963bd5b91c257b3c3d3c1f52cdf2b5f4c9d1a/pubsub/topic.go#L76

With regards to what would make this test fail, it exists to more than anything to statically verify the setting is exposed.

Reply:
What verifies that the settings used came from what's on line 186 instead of the defaults?

Reply:
Given that the change is to the interface there is a limit to how far we can verify this. that limit (without testing google's own pubsub code) is that we can verify the Public variable is exposed in by the struct.

Reply:
I see two other similar PRs to draw potential inspiration from:
1. The original PR adding receiveSettings https://github.com/cloudevents/sdk-go/pull/396/files in which the default value is declared in the sdk go and then a simple check is performed in the test to see if the value is equal to the default.
1. This PR exposing a custom host option https://github.com/cloudevents/sdk-go/pull/1070/files in which a setter function is added and unit tested. 

There is an argument to be made that we should define our own default publish settings like receive although I don't really see how this would make the sdk more intuitive though.

Reply:
my underlying point here remains that the unit test for receive goes arguably less far then this publishSettings test

Reply:
Is there a way to make it so that w/o the parameter the operation works and then with your settings it fails? I could live with that much.

Or if that can't be done (e.g. because we need an instance of google pubsub and we don't have that), then perhaps try to get an error from the call to CreateTopic() that complains about a parameter in your settings. That'll prove your settings made it to the pubsub code.

WDYT?

Reply:
@JamesBLewis  ping